src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -363,8 +363,16 @@ private IByteBuffer ReceiveBuffer(int responseTo, CancellationToken cancellation
                     receiveLockRequest.Task.GetAwaiter().GetResult(); // propagate exceptions
                     while (true)
                     {
-                        var buffer = ReceiveBuffer(cancellationToken);
-                        _dropbox.AddMessage(buffer);
+                        try
+                        {
+                            var buffer = ReceiveBuffer(cancellationToken);
+                            _dropbox.AddMessage(buffer);
+                        }
+                        catch (Exception ex)
+                        {
+                            _dropbox.AddException(responseTo, ex);
```

Comment:
Why add the exception to ONLY `responseTo`?

If the dropbox has multiple waiting messages they should all get the same exception applied to them.

Reply:
This was done by analogy to the logic here: [link](https://github.com/MikalaiMazurenka/mongo-csharp-driver/blob/af8fd4e5ff95592c658227fccd303f83002a1706/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L379)
```
var ignored = messageTask.ContinueWith(
	t => { _dropbox.RemoveMessage(responseTo).Dispose(); },
	TaskContinuationOptions.OnlyOnRanToCompletion);
throw;
```
Where in case of exception only current response buffer is disposed.

Reply:
Yes, but the analogy is inaccurate.

When an exception occurs ALL pending messages in the dropbox are doomed and the exception should be propagated to ALL uncompleted Tasks in the dropbox.

Reply:
So you need to find a thread-safe reliable way to find all uncompleted Tasks in the dropbox and call TrySetException on All of them.

This line of code should be something like:

    _dropbox.AddException(ex);

Reply:
So the `TaskCompletionSource.TrySetException()` ([link](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcompletionsource-1.trysetexception?view=netcore-3.1)) wouldn't affect tasks with 
> RanToCompletion, Faulted, or Canceled

states, and the method is thread-safe ([link](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcompletionsource-1?view=netcore-3.1#:~:text=All%20members%20of%20TaskCompletionSource%3CTResult,used%20from%20multiple%20threads%20concurrently.)):
> All members of TaskCompletionSource<TResult> are thread-safe and may be used from multiple threads concurrently

But I still don't get why we should invalidate all tscs on this line, while buffer is only cleared for one responseTo here: [link](https://github.com/MikalaiMazurenka/mongo-csharp-driver/blob/af8fd4e5ff95592c658227fccd303f83002a1706/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L379)

Reply:
As I commented before I no longer remember why this line exists:

https://github.com/MikalaiMazurenka/mongo-csharp-driver/blob/af8fd4e5ff95592c658227fccd303f83002a1706/src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs#L379

it no longer makes sense to me. If an exception was thrown then there NO message was read and there shouldn't be any buffer to remove.

Reply:
This catch clause wasn't used, so I removed it.